### ISSUE TITLE
Interfaces - carp only interface not functional anymore as of 22.1.1

### ISSUE THREAD

**Issue by AdSchellevis** [opnsense/core#5742]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [x] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Describe the bug**

Refactoring `interfaces_primary_address()` in https://github.com/opnsense/core/commit/65779b80bb1ae72ec49654eef85472a4ecc9a6d7 lead to an unintended side affect where interfaces which have "None" selected as address type will never return an address, even when a CARP address is being asigned.

Although this scenario is less common, it should be possible to only add addresses using (carp) aliases. Primary in this context will always be opaque (first listed in `ifconfig`)

**To Reproduce**

Create an interface, select both ipv4 and ipv6 types as None and add a CARP address for IPv4, next go to (for example) `Services: DHCPv4: Relay` and notice the new interface isn't in the list of valid options although it does have an address.

**Expected behavior**

When no static/dhcp address is configured `interfaces_primary_address()` should return the first address it finds.

**Describe alternatives you considered**

None

**Environment**

Software version used and hardware type if relevant, e.g.:

OPNsense 22.1.1-6 (amd64, OpenSSL).

**Comment by fichtner** [comment]
> Although this scenario is less common, it should be possible to only add addresses using (carp) aliases. Primary in this context will always be opaque (first listed in ifconfig)

If we revert back to this state a lot of work and fixes will be for nothing. ;(

> Create an interface, select both ipv4 and ipv6 types as None and add a CARP address for IPv4

Wasn't this introduced rather recently in 21.7.x?

**Comment by AdSchellevis** [comment]
> If we revert back to this state a lot of work and fixes will be for nothing. ;(

Not necessarily, but first reports are dropping in, likely won't be the last and impact is extremely hard to assess.

> Wasn't this introduced rather recently in 21.7.x?

Not as far as I know, no, I've seen more setups only using CARP addresses, we removed the requirement back in 2018 https://github.com/opnsense/core/commit/075a0b7a4abdc20a379fec6773a208a7effc73ea

**Comment by fichtner** [comment]
But you describe e6f9b9a38832e8 saying None/None is selectable which was added in 22.1 actually.

**Comment by AdSchellevis** [comment]
@fichtner you could already add it the other way around, new interface "None/None" and add a carp address. Assigning an address first and a carp one would lead to the situation where you couldn't remove the "primary" one, which is what https://github.com/opnsense/core/commit/e6f9b9a38832e8c4c3429018691bcadcfa3c1db8 prevented from happening. 

Now both paths lead to the same situation, which was already possible for years.

Using only a carp address happens when you can't spare the additional ones for maintenance, I know more people do that, I'm just not sure how bad the issue is now. I can propose a fix which does lead to the right situation if you like, all of this really is about priority, as we can't trust "first address on interface" which is what you tried to fix, I just missed the fact that it's now not able to return the correct address anymore for these situations.

**Comment by fichtner** [comment]
I'm not even sure it's possible via standard to initiate connection from CARP address. It sounds like a bug in the implementation of the OS: the CARP address can respond, but it can can never initiate a connection by itself. But apparently it works despite all of it.

8ef7cac57075 is then what you mean... We can fudge none/none behaviour in the primary functions, but that's about it. The larger issue is that one VIP is primary but what is the requirement for multiple VIPs. There's only chaos there as before.

**Comment by fichtner** [comment]
Is this solely about https://github.com/opnsense/core/blob/ab71380a078fc7aa8cc07c50421d78f970d26da0/src/etc/inc/plugins.inc.d/dhcpd.inc#L1740 ?

get_interface_ip() can grab the carp address, but it would be nicer if the GUI knew or we emulate it there.

get_interface_ip() isn't used that much anymore. Trying a generic fix two layers below seems like what we had before.

**Comment by AdSchellevis** [comment]
> I'm not even sure it's possible via standard to initiate connection from CARP address. It sounds like a bug in the implementation of the OS: the CARP address can respond, but it can can never initiate a connection by itself. But apparently it works despite all of it.

That's simply not true, when master, it can do both. it's an address like any other, just being "cut off" when in backup.

> We can fudge none/none behaviour in the primary functions, but that's about it. The larger issue is that one VIP is primary but what is the requirement for multiple VIPs. There's only chaos there as before.

We can either figure out where "primary" is a safe concept or make sure we always pick an address. Both could be viable options, the current state is just not right. As mentioned, I haven't invested the full impact, just letting you know if we need an address (any), carp ones are valid and have always been.

**Comment by fichtner** [comment]
Fair enough. However, I'm not convinced the service should gloss over the fact that it wants a CARP address specifically. The services that require this support it well. If we just do opportunistic lookups if we fail nothing really changes. It might be a common way to do it but for the code it can be non-obvious -- it would have been better a long time ago to support a CARP switch in the static interface address if that's what's implied anyway. Sort of why this change came to be to lock out ambiguous behaviour.

The root cause might be https://github.com/opnsense/core/issues/5517#issuecomment-1026774832 which was also introduced in 22.1.1.

Options:

1. Explicitly add support listening to CARP for services that want to do it. Which other services should this patch?
2. Force a CARP address if none was selected. Will it be the first in the list? How will users be able to reorder it? Are normal IP address aliases to be considered equal or not?
3. Allow CARP to be primary again which kills the nobind feature of #5517
